### ISSUE TITLE
Android AudioDispatcher from file

### ISSUE THREAD

**Issue by modularmind** [JorenSix/TarsosDSP#59]
Hi,

Great library!

I'm trying to process audio on Android. I'm building an app to do Voice-to-MIDI transcriptions, but I'm trying to do it from a recorded file (a WAV file made out of an AudioDispatcherFactory.fromDefaultMicrophone and adding a WriterProcessor, this works). The problem is that AudioDispatcherFactory.fromFile is yet to be implemented on Android, so by the time is there another way to create an AudioDispatcher from a file? I'd really appreciate any kind of information to solve my issue!

EDIT: What I'm looking for is a way to do processing on the same recorded signal, so the final user can change some parameters of a chosen algorithm and apply it to the same signal and produce different corresponding results.

Regards,

Carlos.

**Comment by JorenSix** [comment]
Hi, 

Please see the latest version of the Android AudioDispatcherFactory. It now includes a fromPipe() method. You can use together with ffmpeg to decode audio (mp3, ogg,  wav, flac, ape,...) and get an easy to use audio stream. From then on you can use every feature in TarsosDSP available.

For more information see the post on my blog titled [Decode MP3s and other Audio formats the easy way on Android](http://0110.be/posts/Decode_MP3s_and_other_Audio_formats_the_easy_way_on_Android).

**Comment by modularmind** [comment]
Hi. Thank you for your answer. Unfortunately I couldn't try that solution because the latest Android release (2.3) gives me errors. I've spend 4 days trying to make it work without any success. And if I change the library to version 2.2, it simply works.

One of the more common errors is this: 

```
Error:Execution failed for task ':app:preDexDebug'.
> com.android.ide.common.process.ProcessException: org.gradle.process.internal.ExecException: Process 'command '/usr/java/jdk1.8.0_60/bin/java'' finished with non-zero exit value 1
```

I've tried everything: changing JDK version, changing Gradle version, updating everything in the SDK, cleaning and rebuilding the project, minimizing methods, etc. Here's a more detailed output:

```
Information:Gradle tasks [:app:assembleDebug]
:app:preBuild UP-TO-DATE
:app:preDebugBuild UP-TO-DATE
:app:checkDebugManifest
:app:prepareDebugDependencies
:app:compileDebugAidl UP-TO-DATE
:app:compileDebugRenderscript UP-TO-DATE
:app:generateDebugBuildConfig UP-TO-DATE
:app:generateDebugAssets UP-TO-DATE
:app:mergeDebugAssets UP-TO-DATE
:app:generateDebugResValues UP-TO-DATE
:app:generateDebugResources UP-TO-DATE
:app:mergeDebugResources UP-TO-DATE
:app:processDebugManifest UP-TO-DATE
:app:processDebugResources UP-TO-DATE
:app:generateDebugSources UP-TO-DATE
:app:processDebugJavaRes UP-TO-DATE
:app:compileDebugJavaWithJavac UP-TO-DATE
:app:compileDebugNdk UP-TO-DATE
:app:compileDebugSources UP-TO-DATE
:app:preDexDebug
UNEXPECTED TOP-LEVEL EXCEPTION:
java.lang.RuntimeException: Exception parsing classes
	at com.android.dx.command.dexer.Main.processClass(Main.java:752)
	at com.android.dx.command.dexer.Main.processFileBytes(Main.java:718)
	at com.android.dx.command.dexer.Main.access$1200(Main.java:85)
	at com.android.dx.command.dexer.Main$FileBytesConsumer.processFileBytes(Main.java:1645)
	at com.android.dx.cf.direct.ClassPathOpener.processArchive(ClassPathOpener.java:284)
	at com.android.dx.cf.direct.ClassPathOpener.processOne(ClassPathOpener.java:166)
	at com.android.dx.cf.direct.ClassPathOpener.process(ClassPathOpener.java:144)
	at com.android.dx.command.dexer.Main.processOne(Main.java:672)
	at com.android.dx.command.dexer.Main.processAllFiles(Main.java:574)
	at com.android.dx.command.dexer.Main.runMonoDex(Main.java:311)
	at com.android.dx.command.dexer.Main.run(Main.java:277)
	at com.android.dx.command.dexer.Main.main(Main.java:245)
	at com.android.dx.command.Main.main(Main.java:106)
Caused by: com.android.dx.cf.iface.ParseException: bad class file magic (cafebabe) or version (0034.0000)
	at com.android.dx.cf.direct.DirectClassFile.parse0(DirectClassFile.java:472)
	at com.android.dx.cf.direct.DirectClassFile.parse(DirectClassFile.java:406)
	at com.android.dx.cf.direct.DirectClassFile.parseToInterfacesIfNecessary(DirectClassFile.java:388)
	at com.android.dx.cf.direct.DirectClassFile.getMagic(DirectClassFile.java:251)
	at com.android.dx.command.dexer.Main.parseClass(Main.java:764)
	at com.android.dx.command.dexer.Main.access$1500(Main.java:85)
	at com.android.dx.command.dexer.Main$ClassParserTask.call(Main.java:1684)
	at com.android.dx.command.dexer.Main.processClass(Main.java:749)
	... 12 more
1 error; aborting
Error:Execution failed for task ':app:preDexDebug'.
> com.android.ide.common.process.ProcessException: org.gradle.process.internal.ExecException: Process 'command '/usr/java/jdk1.8.0_60/bin/java'' finished with non-zero exit value 1
Information:BUILD FAILED
Information:Total time: 1.931 secs
Information:1 error
Information:0 warnings
Information:See complete output in console
```

**Comment by JorenSix** [comment]
Hi,

I'm not sure of the exact nature of the error but it seems to be related to the source level of the compiled class files: `bad class file magic (cafebabe) or version (0034.0000)`. 

If I set the source level to 1.7 the error disappears. I have fixed the ant build file to set the source level to 1.7 and have uploaded new jar-files for the [2.3 release](http://0110.be/releases/TarsosDSP/TarsosDSP-2.3/]). 

Let me know if this fixes the problem.
